Framework/Muon/src/MuonPairingAsymmetry.cpp

```diff
@@ -0,0 +1,476 @@
+#include "MantidAPI/AlgorithmManager.h"
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidAPI/WorkspaceGroup.h"
+#include "MantidDataObjects/TableWorkspace.h"
+#include "MantidHistogramData/HistogramMath.h"
+#include "MantidKernel/ArrayProperty.h"
+#include "MantidKernel/EnabledWhenProperty.h"
+#include "MantidKernel/MandatoryValidator.h"
+#include "MantidMuon/MuonAlgorithmHelper.h"
+
+using namespace Mantid::API;
+using namespace Mantid::DataObjects;
+using namespace Mantid::Kernel;
+
+namespace {
+bool checkPeriodInWorkspaceGroup(const int &period,
+                                 WorkspaceGroup_sptr workspace) {
+  return period <= workspace->getNumberOfEntries();
+}
+
+bool isMultiPeriod(Workspace_sptr ws) {
+  if (auto tmp = boost::dynamic_pointer_cast<WorkspaceGroup>(ws)) {
+    return true;
+  }
+  return false;
+}
+
+int countPeriods(Workspace_sptr ws) {
+  auto tmp = boost::dynamic_pointer_cast<WorkspaceGroup>(ws);
+  return tmp->getNumberOfEntries();
+}
+
+bool checkConsistentPeriods(Workspace_sptr ws1, Workspace_sptr ws2) {
+  if (isMultiPeriod(ws1) && isMultiPeriod(ws2)) {
+    if (countPeriods(ws1) != countPeriods(ws2)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+MatrixWorkspace_sptr getWorkspace(WorkspaceGroup_sptr group, const int &index) {
+  auto ws = group->getItem(index);
+  return boost::dynamic_pointer_cast<MatrixWorkspace>(ws);
+}
+
+bool is_alnum_underscore(char c) {
+  return (isalpha(c) || isdigit(c) || (c == '_'));
+}
+
+MatrixWorkspace_sptr groupDetectors(MatrixWorkspace_sptr workspace,
+                                    const std::vector<int> &detectorIDs) {
+
+  auto outputWS = WorkspaceFactory::Instance().create(workspace, 1);
+
+  std::vector<size_t> wsIndices =
+      workspace->getIndicesFromDetectorIDs(detectorIDs);
+
+  if (wsIndices.size() != detectorIDs.size())
+    throw std::invalid_argument("Some of the detector IDs were not found");
+
+  outputWS->getSpectrum(0).clearDetectorIDs();
+  outputWS->setSharedX(0, workspace->sharedX(wsIndices.front()));
+
+  auto hist = outputWS->histogram(0);
+  for (auto &wsIndex : wsIndices) {
+    hist += workspace->histogram(wsIndex);
+    outputWS->getSpectrum(0).addDetectorIDs(
+        workspace->getSpectrum(wsIndex).getDetectorIDs());
+  }
+  outputWS->setHistogram(0, hist);
+  outputWS->getSpectrum(0).setSpectrumNo(static_cast<int32_t>(1));
+  return outputWS;
+}
+
+// Convert a Workspace_sptr (which may be single period, MatrixWorkspace, or
+// multi period WorkspaceGroup) to a WorkspaceGroup_sptr
+WorkspaceGroup_sptr workspaceToWorkspaceGroup(Workspace_sptr workspace) {
+
+  WorkspaceGroup_sptr ws1;
+  if (isMultiPeriod(workspace)) {
+    ws1 = boost::dynamic_pointer_cast<WorkspaceGroup>(workspace);
+  } else {
+    ws1 = boost::make_shared<WorkspaceGroup>();
+    ws1->addWorkspace(boost::dynamic_pointer_cast<MatrixWorkspace>(workspace));
+  }
+  return ws1;
+}
+
+} // namespace
+
+namespace Mantid {
+namespace Muon {
+
```

Comment:
should the above be in a different file as it is not part of the algorithm?

Reply:
I would argue they are part of the algorithm just factored out into an anonymous namespace at the top. This was presumably done to make refactoring easier later if they were found to be commonly used elsewhere. This does mean it should be easy to move them to a different file if you want to though.